<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=gbk">
    <title>kpLogo: k-mer probability logo</title>
    <style>
input[type="submit"]{
/* change these properties to whatever you want */
background-color: #555;
color: #fff;
border-radius: 10px;
font-size:1em;
height:30px;

}

code {
color: SteelBlue  ;
font-size:0.9em;
}


h2 {
font-family: Verdana;
color: SteelBlue  ;
}
h3 {
font-family: Verdana;
color: SteelBlue  ;
}
h4 {
font-family: Verdana;
color: SteelBlue  ;
}

td {
    vertical-align: top;
}

a:link {
    text-decoration: none;
	color: SlateBlue;
}

  </style>
    <style>
input[type="file"]{
/* change these properties to whatever you want */
background-color: #555;
color: #fff;
border-radius: 10px;
font-size:1em;
}
  </style>
    <style>
input[type="text"]{
/* change these properties to whatever you want */
background-color: #aaa;
color: #fff;
}
  </style>
  </head>
  <body style="margin:0 auto; width: 970px; text-align: center">
    <div style="font-family: Helvetica,Arial,sans-serif;text-align:
      left;" id="title"> <br>
      <h1><span class="title"><i>k</i>pLogo: <i>k</i>-mer probability logo</span><br>
      </h1>
    </div>
    <div style=" border-top: 8px solid DimGray; width: 970px"></div>
    <div style=" border-top: 8px solid White; width: 970px"></div>
<table style="text-align: left; width: 436px; height: 32px; margin-left: auto; margin-right: 0px;" border="0" cellpadding="0" cellspacing="0">
  <tbody>
    <tr>
      <td style="vertical-align: top; width: 69px;">
      <form action="./index.html"> <input value="Home" type="submit" style="color: GreenYellow" ></form>
      </td>
      <td style="vertical-align: top; width: 69px;">
      <form action="./submit.html"> <input value="Server" type="submit"></form>
      </td>
      <td style="vertical-align: top; width: 0px;">
        <form action="./manual.html#install-kpLogo-locally"> <input value="Code" type="submit"></form>
        </td>
        <td style="vertical-align: top; width: 69px;">
        <form action="./manual.html"> <input value="Manual" type="submit"></form>
        </td>
        <td style="vertical-align: top; width: 69px;">
        <form action="./manual.html#examples"> <input value="Examples" type="submit"></form>
      </td>
	  <td style="vertical-align: top; width: 0px;">
	  <form action="./feedback.html"> <input value="Feedback" type="submit"></form>
	  </td>
    </tr>
  </tbody>
</table>

    <div id="TOC" style="font-family: Helvetica,Arial,sans-serif;
      text-align: left;"><span style="font-weight: bold;"></span><br>
      <ul id="mozToc">
        <!--mozToc h1 1 h2 2 h3 3 h4 4 h5 5 h6 6-->
        <li><a href="#What-is-kpLogo">What is <i>k</i>pLogo</a>		</li>
		

		  
        <li><a href="#using-kpLogo">Using <i>k</i>pLogo</a>
          <ul>
            <li><a href="#web-server">Web server</a>
				<ul>
                	<li><a href="#email">Email and data security</a></li>
                	<li><a href="#server-limit">Limitations</a></li>				
				</ul>
                <li><a href="#input">Input</a></li>
				<ul>
                	<li><a href="#sequence-type">Sequence type: DNA/RNA-Protein-Other</a></li>
                	<li><a href="#file-format">File format: Tabular/FASTA</a></li>				
				    <li><a href="#subregion">Sub-region</a></li>
				</ul>
                <li><a href="#positional-kmer"> <i>k</i>-mer counting</a></li>
                <ul>
                	<li><a href="#kmer-length"><i>k</i>-mer length</a></li>
                	<li><a href="#degenerate-residual">Degenerate residue (DNA/RNA)</a></li>				
				    <li><a href="#positional-shift">Positional shift (experimental)</a></li>
				    <li><a href="#filter-by-kmer">Filter by positional <i>k</i>-mer</a></li>
				</ul>
                <li><a href="#statistical-models"> Statistical models</a>
                  <ul>
                	<li><a href="#unweighted">Unweighted (default): binomial test</a></li>
                	<li><a href="#weighted">Weighted: two-sample Student's t-test</a></li>				
				    <li><a href="#ranked">Ranked: Mann-Whitney U test</a></li>
				    <li><a href="#other-parameters">Other parameters</a></li>                  </ul>
                </li>
                <li><a href="#output"> Output</a>
				                 <ul>
                	<li><a href="#logo-plots">Logo plots</a>                	</li>
                	<li><a href="#tabular-data-file">Tabular data file</a></li>				
                 </ul>
                </li>
	            <li><a href="#run-time">Run time and input limit</a>
	            </li>
	            <li><a href="#galaxy">Galaxy Tool Shed</a>
	            </li>
            <li><a href="#install-kpLogo-locally">Install <i>k</i>pLogo locally</a>
            </li>
            <li><a href="#command-line-options">Command Line Options</a></li>
          </ul>
        </li>
        
                <li><a href="#examples">Examples</a>
          <ul>
            <li><a href="#example1">Unweighted: sequence features defining human microRNA genes</a></li>
            <li><a href="#example2">Weighted/Ranked: CRISPR/Cas9 guide RNA efficiency </a></li>
          </ul>
        </li>
        <li><a href="#copyright">Copyright</a></li>
        <li><a href="#author">Author</a></li>
        <li><a href="#acknowlegements">Acknowledgements</a></li>
        <li><a href="#reference">Reference</a></li>
      </ul>
      <br>
    </div>
    <div style="font-family: Helvetica,Arial,sans-serif; text-align:
      left;">
      <h1><a class="mozTocH1" name="What-is-kpLogo"></a>What is <i>k</i>pLogo</h1>
	  
	  
	  
	  
      <font color="Blue"><i>k</i>-mer probability logo (<i>k</i>pLogo)</font> is a probability-based logo tool for integrated detection and visualization of position-specific ultra-short motifs from a set of aligned sequences. <i>k</i>pLogo also overcomes the limitations of conventional motif-visualization tools in handling positional interdependencies and utilizing ranked or weighted sequences increasingly available from high-throughput assays. 
	  
	  <p>
	  Given a set of sequences of identical length, either unweighted, weighted, or ranked, <i>k</i>pLogo enumerates all possible <i>k</i>-mers of user-specified lengths, evaluates their presence at each position in all input sequences, and reports their enrichment and depletion at each position as determined using an appropriate statistical model. To visualize the results, kpLogo generates a unique type of logo plot called the <i>k</i>-mer logo, in which at each position the most significant <i>k</i>-mer is plotted vertically with the total height scaled to its <i>P</i> value (log10 transformed) or test statistics, as appropriate. In addition to <i>k</i>-mer logo, <i>k</i>pLogo also generates logo plots for monomer frequency and information content.
	  <p>
	  
	  The diagram below highlights two overlapping CNNC motifs at specific positions within primary transcripts of human microRNAs, which are important for microRNA processing (see <a href="#examples">Examples</a>):
	  
	  				<p style="text-align:center;"><img src="./diagram.jpg" width="70%"></p>
				
	      
  	</p>
	  
		<p><b>Limitations of current motif tools</b></p>
	  
	  Existing motif discovery tools and visualization tools have fundamental limitations, but there are also strong synergy between them: 
	  
	  <p>
		  Motif visualization (sequence logos): 1) cannot handle weighted or ranked sequences; 2) models positions with single-nucleotide resolution but lack the ability to model and detect positional interdependence in motifs. 
	  </p>
	  <p> Motif discovery: handles weighted or ranked sequences and models positional interdependence, but typically ignores positional information and thus misses ultra-short motifs (with lengths 1-4 letters) or other information-poor motifs whose specificities are conferred by both sequence identity and relative position.
		  
	  <p><b>Unique features of <i>k</i>pLogo </b></p>
      <i>k</i>pLogo combines motif discovery with visualization. Compared to
      other tools that generate sequence logos, <i>k</i>pLogo is unique in several ways:<br> <br>
      1. handles ranked or weighted sequences<br>
      2. performs statistical tests to discover short motifs <br>
	  3. visualizes not only single letters but also short motifs using <i>k</i>-mer logo<br>
      4. plots frequency logo, information content logo, probability logo, and <i>k</i>-mer logo in one run<br>
      5. allows degenerate letters (DNA/RNA) and gapped motifs<br>
      6. logos can be generated with colorblind friendly colors<br>
	  
		
<h1><a class="mozTocH1" name="using-kpLogo"></a>Using <i>k</i>pLogo</h1>

      <i>k</i>pLogo can be accessed as a webserver: <a href="http://kplogo.wi.mit.edu">http://kplogo.wi.mit.edu</a>. In addition, the C++ source code and the web code can be downloaded and installed locally. <br>
	  
<h2><a class="mozTocH2" name="web-server"></a>Web server</h2>
	To use <i>k</i>pLogo online, please visit <a href="http://kplogo.wi.mit.edu/">http://kplogo.wi.mit.edu</a><br>
	  
	<p style="text-align:center;"><img src="./manual/main.png" width="85%"></p>
	  
	<p>Users paste or upload sequences to the server, choose sequence type (DNA/RNA, protein, or other) and analysis type (unweighted, weighted, or ranked), then submit the job using default setting. Once submitted, users are redirected to the result page which refreshes every second until the job is finished.
	  
<h3><a class="mozTocH3" name="email"></a>Email and data security</h3>
	  
	  <p>Note that user data will be <span style="color: red;">removed 10 days</span> after the job is submitted. Please download the data as soon as it is
      available. If the user wishes, all data related to the job can be  deleted immediately using  a button in the result page. No copy of the data will be stored or used for other purposes. </p>
	  
	  Users are recommended to provide an optional email address to receive notifications. If provided, two emails will be sent: one immediately after
      submitting the job, and the other after results are available. Both emails contain a link to the result page. The second email contains a PDF file with all logo plots. Note that users need to <span style="color: red;">keep the result page open</span> in a web browser for the second email to be sent. If the result page is closed before the job is finished, the second email will not be sent but users will still be able to open the result page using the link in the first email.<br>
      <br>
    
<h3><a class="mozTocH3" name="server-limit"></a>Limitations</h3>

      The web server implements most functionality of <i>k</i>pLogo, and should be sufficient for most users. However, for simplicity and to avoid excess demand of server resources, a few restrictions have been imposed on the web version of <i>k</i>pLogo:<br>
      <ol>
          <li><span style="font-weight: bold;">Input file size</span>: Each input file should not exceed 2MB in size.<br> </li>
          <li><span style="font-weight: bold;"><i>k</i>-mer length</span>: the max length of <i>k</i>-mer is limited to 6 for RNA/DNA sequences, and 3 for protein and other sequences.<br></li>
          <li><span style="font-weight: bold;">sequence shuffling</span>: if choosing to shuffle input sequences to generate the background, one can only preserve up to 4-residue frequency, and for each sequence, shuffle no more than 1000 times. <br> </li>
        <li><span style="font-weight: bold;">model prediction</span>: the binary version allows the user to score a new set of sequences using the positional sequence specificity model learned from the input. The web version does not support this functionality at this moment.<br>
        </li>
      </ol>
	  
     <h2><a class="mozTocH2" name="input"></a>Input</h2>
      <h3><a class="mozTocH3" name="sequence-type"></a>Sequence type: DNA/RNA/protein/other</h3>
	  
	  <p style="text-align:center;"><img src="./manual/sequence-type.png" width="45%"></p>
	  
		<i>k</i>pLogo is designed primarily for analyzing DNA or RNA sequences, but can also be used to study protein and other arbitrary types of sequences. To run on other types of sequences, choose "Other" in "Sequence type" dropdown menu, and in the text box enter all allowed residues/letters:
		
  	  <p style="text-align:center;"><img src="./manual/sequence-type-other.png" width="30%"></p>
		
	  <p><b>Command-line options: </b>When <code>-alphabet</code> is not specified by users, <i>k</i>pLogo uses the default <code>-alphabet ACGT</code>. Note RNA will be converted to DNA by default so for RNA input there is no need to specify <code>-alphabet ACGU</code>.	   To run <i>k</i>pLogo on protein sequences, please use <code>-alphabet protein</code>, which is equivalent to <code>-alphabet ACDEFGHIKLMNPQRSTVWY</code>.  Similarly, to run <i>k</i>pLogo on other types of sequences, use <code>-alphabet </code> to specify all allowed residues.</p>

      <h3><a class="mozTocH2" name="file-format"></a>File format</h3>
	  
  	  <p style="text-align:center;"><img src="./manual/file-format.png" width="50%"></p>
	  
	  <i>k</i>pLogo supports three types of input sequences with respect to the weight of individual sequence: unweighted, ranked, or weighted. Weighted sequences must be provided in a tabular/column file, with sequences and numeric weights in separate columns. For unweighted or ranked sequences, both FASTA and tabular files can be used. For ranked sequences, the sequences must be ranked with more important sequences appearing earlier in the file. <font color="red">Note that all sequences will be converted to upper case by default</font>. Lines starting with <font color="blue">  # </font> will be treated as comments and be ignored.
	  
	  <p><b>Tabular / Column file</b></p>
      <a><i>k</i>pLogo will treat <font color="red">any</font> of the following characters as delimiter: tab ("\t"), comma (","), vertical bar ("|"), and space (" ").
        <br>
        <br>
        Example 1: minimal ranked or unweighted tabular input <br>
        <br>
      </a>
      <div style="border: 0px solid ; padding: 10px 40px; background:
        rgb(221, 221, 221) none repeat scroll 0% 50%;
        -moz-background-clip: -moz-initial; -moz-background-origin:
        -moz-initial; -moz-background-inline-policy: -moz-initial;
        font-family: Monaco;"><a><small>CGTAGCTAGCTACGTAGCTA<br>
            CGTACGTAGCTAGCTAGCTA<br>
            GCTGTGATCGTAGCTAGCTG<br>
            ...</small></a></div>
      <a><br>
        Example 2: minimal weighted input. Note: no need to sort sequences by weight <br>
        <br>
      </a>
      <div style="border: 0px solid ; padding: 10px 40px; background:
        rgb(221, 221, 221) none repeat scroll 0% 50%;
        -moz-background-clip: -moz-initial; -moz-background-origin:
        -moz-initial; -moz-background-inline-policy: -moz-initial;
        font-family: Monaco;"><a><small>
			CGTAGCTAGCTACGTAGCTA 3.4<br>
            CGTACGTAGCTAGCTAGCTA 8.7<br>
            GCTGTGATCGTAGCTAGCTG 1.2<br>
            ...</small></a></div>
      <br>
      <a>Example 3: additional columns in input.  <br>
      </a><a><br>
      </a>
      <div style="border: 0px solid ; padding: 10px 40px; background:
        rgb(221, 221, 221) none repeat scroll 0% 50%;
        -moz-background-clip: -moz-initial; -moz-background-origin:
        -moz-initial; -moz-background-inline-policy: -moz-initial;
        font-family: Monaco;"><a><small>
			seq1 CGTAGCTAGCTACGTAGCTA 3.4<br>
            seq2 CGTACGGAGCTAGCTAGCTA 8.7<br>
            seq3 GCTGTGAGCGTAGCTAGCTG 1.2<br>
            ...</small></a></div>
      <a><br>
		  
		  In cases like this where the sequences are not in column 1, and / or the weights are not in column 2, then one needs to specify the columns by chaning the numbers in those two boxes:
		  
	  	  <p style="text-align:center;"><img src="./manual/seq-weight-column.png" width="50%"></p>
		  
		  <i>Command-line options</i>: <code>-seq 2 -weight  3</code><br>
      <br>
		  <p><b>FASTA</b></p>
      <a>Here is an example of a FASTA file: <br>
      </a><a><br>
      </a>
      <div style="border: 0px solid ; padding: 10px 40px; background:
        rgb(221, 221, 221) none repeat scroll 0% 50%;
        -moz-background-clip: -moz-initial; -moz-background-origin:
        -moz-initial; -moz-background-inline-policy: -moz-initial;
        font-family: Monaco;"><a><small>&gt;seq1
            <br>
            CGTAGCTAGCTAGCTAGCTA<br>
            &gt;seq2<br>
            CGTAGCTAGCTAGCTAGCTA<br>
            &gt;seq3<br>
            CGTAGCTAGCTAGCTAGCTA<br>
            &gt;seq4<br>
            CGTAGCTAGCTAGCTAGCTA<br>
            ...</small></a></div>
      <a><br>
        Notes on FASTA file format: <br>
        1. Each sequence includes an ID line and a sequence line <br>
        2. IDs always start with '&gt;' <br>
        3. IDs are not used by <i>k</i>pLogo and thus do not need to be unique <br>
        4. Sequences do not need to be of the same length (see option
        <code>-region</code>)<br>
        5. Can be treated as ranked (see option <code>-ranked</code>) but not
        weighted (need to use tabular format) </a>

      <h3><a class="mozTocH3" name="subregion"></a>Analyze a sub-region</h3>
	  To analyze only part of the positions, specify the start and end in the textboxes:
	  
  	  <p style="text-align:center;"><img src="./manual/subregion.png" width="80%"></p>
	  
	  Examples: <br>
	  <b>start=1,end=0</b>: analyze all positions without trimming. All input sequences need to be of identical length. <br>
	  <b>start=1,end=10</b>: analyze the first 10 positions. Input sequences need not to be of identical length. <br>
	  <b>start=-9,end=0</b>: analyze the last 10 positions. Input sequences need not to be of identical length. <br>
	  <b>start=1,end=-9</b>: analyze all but the last 10 positions, i.e. trim 10 residues from the end. Input sequences need to be of identical length. <br>
	  
	  
	   <p><b>Command-line options</b>: <code>-region start,end</code></p>
	   
	   
	   
      <h2><a class="mozTocH2" name="positional-kmer"></a>Positional <i>k</i>-mer with shift and degeneracy</h2>

		<p> A positional <i>k</i>-mer (or simply <i>k</i>-mer in this document) is a short sequence of length <i>k</i> associated with a specific position. Throughout this document, a positional <i>k</i>-mer has the format of <code>sequence:position:shift</code>, or simply <code>sequence:position</code> if no shift is allowed (default). For example, most introns start with a GU dinucleotide, and this dinucleotide motif (<i>k</i> = 2) can be represented as <code>GU:1:0</code>, i.e. GU at position 1 with no shift. The position of the first residue is used as the position of the <i>k</i>-mer.		</p>
		<h3><a class="mozTocH3" name="kmer-length"></a><i>k</i>-mer length</h3>
		
		<i>k</i>-mers up to 6 DNA/RNA residues or 3 protein residues are supported in the webserver, with the default being 1 to 4 for DNA/RNA and 1 to 2 for protein and other sequences. The command-line tool allows more flexibility in <i>k</i>-mer length. Please note that the number of all possible  <i>k</i>-mers increases exponentially with length and thus has a large influence on the run time. 
        <p style="text-align:center;"><img src="./manual/kmer-length.png" width="65%"></p>



      <h3><a class="mozTocH3" name="degenerate-residual"></a>Degenerate
        residue (DNA/RNA)</h3>
		
		<p>Biological sequence signals can  tolerate some mismatches, leading to degeneracy in the motif. For example, in addition to AAUAAA, AUUAAA is also used as a polyadenylation signal. Therefore a more general form of the polyadenylation signal is <code>AWUAAA</code>, where W is either A or U. <i>k</i>pLogo supports all degenerate codes defined in the <a
        href="http://en.wikipedia.org/wiki/Nucleic_acid_notation">IUPAC code</a> for nucleic acids (see table below). </p>
        
                  	  <p style="text-align:center;"><img src="./manual/degenerate.png" width="55%"></p>

        
		<p>The default of the webserver is to allow N (any of A/C/G/T/U) in the middle of a <i>k</i>-mer, which represents a gapped <i>k</i>-mer (<code>-degenerate ACGTN</code> or simply <code>-gapped</code>). Users can choose to not include degenerate residues (default of command-line), use all IUPAC-defined residues (<code>-degenerate all</code>), or only allow a subset of IUPAC residues by choosing 'Other' in the dropdown menu.</p>
        
          	  
          	            	  <p style="text-align:center;"><img src="./manual/degenerate-other.png" width="50%"></p>
	  <p>For example, to allow only purine (R = A or G) and pyrimidine (Y = C or T), enter <code>ACGTRY</code> in the text box, <span style="color: red;">including
      non-degenerate residues</span>.<br>
      </p>
	  <table style="width: 445px; height: 470px;" border="1"
        cellpadding="2" cellspacing="2" style="text-align:center;" >
        <tbody>
          <tr>
            <td style="font-weight: bold; width: 60px; text-align:
              center;">Symbol </td>
            <td style="font-weight: bold; width: 200px; text-align:
              center;">Description </td>
            <td style="font-weight: bold; width: 25px; text-align:
              center;" colspan="5" rowspan="1">Bases represented</td>
          </tr>
          <tr>
            <td style="width: 60px; text-align: center;">A </td>
            <td style="width: 200px; text-align: center;">Adenine </td>
            <td style="width: 25px; text-align: center;">A </td>
            <td style="width: 25px; text-align: center;"><br>
            </td>
            <td style="vertical-align: top; width: 25px; text-align:
              center;"><br>
            </td>
            <td style="vertical-align: top; text-align: center;"><br>
            </td>
            <td colspan="1" rowspan="5" style="vertical-align: middle;
              text-align: center;">1<br>
            </td>
          </tr>
          <tr>
            <td style="width: 60px; text-align: center;">C </td>
            <td style="width: 200px; text-align: center;">Cytosine </td>
            <td style="width: 25px; text-align: center;"><br>
            </td>
            <td style="vertical-align: top; width: 25px; text-align:
              center;">C<br>
            </td>
            <td style="vertical-align: top; width: 25px; text-align:
              center;"><br>
            </td>
            <td style="vertical-align: top; text-align: center;"><br>
            </td>
          </tr>
          <tr>
            <td style="width: 60px; text-align: center;">G </td>
            <td style="width: 200px; text-align: center;">Guanine </td>
            <td style="width: 25px; text-align: center;"><br>
            </td>
            <td style="vertical-align: top; width: 25px; text-align:
              center;"><br>
            </td>
            <td style="vertical-align: top; width: 25px; text-align:
              center;">G<br>
            </td>
            <td style="vertical-align: top; text-align: center;"><br>
            </td>
          </tr>
          <tr>
            <td style="width: 60px; text-align: center;">T </td>
            <td style="width: 200px; text-align: center;">Thymine </td>
            <td style="width: 25px; text-align: center;"><br>
            </td>
            <td style="vertical-align: top; width: 25px; text-align:
              center;"><br>
            </td>
            <td style="vertical-align: top; width: 25px; text-align:
              center;"><br>
            </td>
            <td style="vertical-align: top; text-align: center;">T<br>
            </td>
          </tr>
          <tr>
            <td style="width: 60px; text-align: center;">U </td>
            <td style="width: 200px; text-align: center;">Uracil </td>
            <td style="width: 25px; text-align: center;"><br>
            </td>
            <td style="vertical-align: top; width: 25px; text-align:
              center;"><br>
            </td>
            <td style="vertical-align: top; width: 25px; text-align:
              center;"><br>
            </td>
            <td style="vertical-align: top; text-align: center;">U<br>
            </td>
          </tr>
          <tr>
            <td style="width: 60px; text-align: center;">W </td>
            <td style="width: 200px; text-align: center;">Weak </td>
            <td style="width: 25px; text-align: center;">A </td>
            <td style="width: 25px; text-align: center;"> <br>
            </td>
            <td style="width: 25px; text-align: center;"><br>
            </td>
            <td style="vertical-align: top; text-align: center;">T<br>
            </td>
            <td colspan="1" rowspan="6" style="vertical-align: middle;
              text-align: center;">2<br>
            </td>
          </tr>
          <tr>
            <td style="width: 60px; text-align: center;">S </td>
            <td style="width: 200px; text-align: center;">Strong </td>
            <td style="width: 25px; text-align: center;"> <br>
            </td>
            <td style="width: 25px; text-align: center;">C<br>
            </td>
            <td style="vertical-align: top; width: 25px; text-align:
              center;">G<br>
            </td>
            <td style="vertical-align: top; text-align: center;"><br>
            </td>
          </tr>
          <tr>
            <td style="width: 60px; text-align: center;">M </td>
            <td style="width: 200px; text-align: center;">aMino </td>
            <td style="width: 25px; text-align: center;">A </td>
            <td style="width: 25px; text-align: center;">C</td>
            <td style="vertical-align: top; width: 25px; text-align:
              center;"><br>
            </td>
            <td style="vertical-align: top; text-align: center;"><br>
            </td>
          </tr>
          <tr>
            <td style="width: 60px; text-align: center;">K </td>
            <td style="width: 200px; text-align: center;">Keto </td>
            <td style="width: 25px; text-align: center;"> <br>
            </td>
            <td style="width: 25px; text-align: center;"><br>
            </td>
            <td style="vertical-align: top; width: 25px; text-align:
              center;">G<br>
            </td>
            <td style="vertical-align: top; text-align: center;">T<br>
            </td>
          </tr>
          <tr>
            <td style="width: 60px; text-align: center;">R </td>
            <td style="width: 200px; text-align: center;">puRine </td>
            <td style="width: 25px; text-align: center;">A </td>
            <td style="width: 25px; text-align: center;"><br>
            </td>
            <td style="vertical-align: top; width: 25px; text-align:
              center;">G<br>
            </td>
            <td style="vertical-align: top; text-align: center;"><br>
            </td>
          </tr>
          <tr>
            <td style="width: 60px; text-align: center;">Y </td>
            <td style="width: 200px; text-align: center;">pYrimidine </td>
            <td style="width: 25px; text-align: center;"> <br>
            </td>
            <td style="width: 25px; text-align: center;">C<br>
            </td>
            <td style="vertical-align: top; width: 25px; text-align:
              center;"><br>
            </td>
            <td style="vertical-align: top; text-align: center;">T<br>
            </td>
          </tr>
          <tr>
            <td style="width: 60px; text-align: center;">B </td>
            <td style="width: 200px; text-align: center;">not A (B comes
              after A)<br>
            </td>
            <td style="width: 25px; text-align: center;"> <br>
            </td>
            <td style="width: 25px; text-align: center;"> C<br>
            </td>
            <td style="width: 25px; text-align: center;"> G<br>
            </td>
            <td style="text-align: center;"> T<br>
            </td>
            <td style="text-align: center; vertical-align: middle;"
              colspan="1" rowspan="4">3</td>
          </tr>
          <tr>
            <td style="width: 60px; text-align: center;">D </td>
            <td style="width: 200px; text-align: center;">not C (D comes
              after C) <br>
            </td>
            <td style="width: 25px; text-align: center;"> A<br>
            </td>
            <td style="width: 25px; text-align: center;"> <br>
            </td>
            <td style="width: 25px; text-align: center;"> G<br>
            </td>
            <td style="text-align: center;"> T<br>
            </td>
          </tr>
          <tr>
            <td style="width: 60px; text-align: center;">H </td>
            <td style="width: 200px; text-align: center;">not G (H comes
              after G) <br>
            </td>
            <td style="width: 25px; text-align: center;"> A<br>
            </td>
            <td style="width: 25px; text-align: center;"> C<br>
            </td>
            <td style="width: 25px; text-align: center;"> <br>
            </td>
            <td style="text-align: center;"> T<br>
            </td>
          </tr>
          <tr>
            <td style="width: 60px; text-align: center;">V </td>
            <td style="width: 200px; text-align: center;">not T (V comes
              after T and U) <br>
            </td>
            <td style="width: 25px; text-align: center;"> A<br>
            </td>
            <td style="width: 25px; text-align: center;"> C<br>
            </td>
            <td style="width: 25px; text-align: center;"> G<br>
            </td>
            <td style="text-align: center;"> <br>
            </td>
          </tr>
          <tr>
            <td style="width: 60px; text-align: center;">N </td>
            <td style="width: 200px; text-align: center;">No idea <br>
            </td>
            <td style="width: 25px; text-align: center;"> A<br>
            </td>
            <td style="width: 25px; text-align: center;"> C<br>
            </td>
            <td style="width: 25px; text-align: center;"> G<br>
            </td>
            <td style="text-align: center;"> T<br>
            </td>
            <td style="text-align: center;">4 <br>
            </td>
          </tr>
        </tbody>
      </table>
      <br>
  


		
        <h3><a class="mozTocH3" name="positional-shift"></a>Positional shift (<font color=Red>experimental</font>)</h3>

      <p>In many  cases, sequence motifs are well positioned but can tolerate some offset. Such positioning flexibility is modeled by the positional shift in <i>k</i>pLogo. For example, the canonical polyadenylation signal, AAUAAA, typically locates about 18 to 23 nucleotides upstream of the cleavage site, can be represented as <code>AAUAAA:-23:5</code>, i.e. the hexamer motif can be found at position -23, allowing a shift of 5, which means the motif can also occur at 5 other positions: -22, -21, -20, -19, and -18. In <i>k</i>pLogo, the position of a <i>k</i>-mer is defined as the position of the first residue in the <i>k</i>-mer, and positional shift is only allowed downstream of the start site, i.e. shift to the right. Positional shift is an experimental feature still under development. For example, currently it is not fully supported by some background models (See below).   </p>
      <p>By default no positional shift is allowed (shift = 0). In the webserver users can choose to use a shift up to 6. The command-line tool allows  larger shifts. </p>
        <p style="text-align:center;"><img src="./manual/shift.png" width="70%"></p>

		
     
      <h3><a class="mozTocH3" name="filter-by-kmer"></a>Filter by positional <i>k</i>-mer</h3>
        <p> <i>k</i>pLogo allows user to filter input sequences by the presence (<code>-select</code>) or absence (<code>-remove</code>) of a list of positional <i>k</i>-mers. This feature enables one to study dependence between positional <i>k</i>-mers.</p>
        
                <p style="text-align:center;"><img src="./manual/kmer-filter.png" width="85%"></p>

      <p>A positional <i>k</i>-mer is specified in the form of <strong>sequence:position:shift</strong>, or <strong>sequence:position</strong> using a default shift of 0. For example, GT:3:0 or GT:3 means GT starting at position 3 with no shift. Note that the sequence can contain degenerate residues. Position coordinates start at 1
      for the first residue in each sequence. </p>
        <p> To filter by multiple positional <i>k</i>-mers, separate each by a comma (,). For example, GT:3:0,ACT:11:0,GGG:45:0 (<font color="red">no space!</font>). Note that when multiple positional <i>k</i>-mers are specified, the presence of any positional <i>k</i>-mer is defined as a match.  </p>
        <p>Here are two more examples: </p>
<p><code>-select G:35:3,CNC:47:0</code> </p>
<p> Select (keep) sequences with G at position 35, 36, 37, or 38 (note shift = 3), <strong>OR</strong> with CNC at position 47. Here N means any nucleotide, thus can be any of CAC, CGC, CTC, and CCC. This will remove sequences that contain none of these positional <i>k</i>-mers.</p>
<p><code>-remove G:35:0,CNC:47:0</code>
<p> Remove sequences containing either G at position 35, or CNC at position 47. Sequences that contain neither of these two <i>k</i>-mers will be analyzed by <i>k</i>pLogo.      </p>

<h2><a class="mozTocH2" name="statistical-models"></a>Statistical
        models</h2>

      <p style="text-align:center;"><img src="./manual/model.png" width="80%"></p>

		<p>For a given value of <i>k</i>, <i>k</i>pLogo enumerates all possible <i>k</i>-mers (allowing degenerate residues when specified), determines their presence / absence at every possible positions in all input sequences, and reports their enrichment / depletion using statistical models specified by users.</p>
       <h4><a class="mozTocH4" name="unweighted"></a>Unweighted (default): binomial test</h4>
     
      When neither <code>-ranked</code> nor <code>-weighted</code> is specified, <i>k</i>pLogo assumes the input is unweighted. <i>k</i>pLogo uses <code>one-sided binomial tests</code> to determine the significance of a <i>k</i>-mer at a position. Specifically, the number of trials is the total number of sequences, and the number of success is the number of sequences containing this <i>k</i>-mer at this position. The hypothesized probability of successes, i.e. the expected / background probability of a sequence containing this <i>k</i>-mer at this position, can be calculated in several ways as specified by users (see details below). A z-score is also reported in the output, assuming normal approximation. The sign of the z-score also indicates enrichment or depletion.<br>
      <br>
      The hypothesized probability of success can be calculated in five different ways:<br>
      <p>
      <b>Background
        sequences.</b> Users can provide a set of background sequences (<code>-bgfile</code>)
      that can be aligned in the same way. In this case, the expected
      probability of a sequence with a particular <i>k</i>-mer at a particular
      position is estimated by the fraction of background sequences with
      this <i>k</i>-mer at this position.         
      <p>
      The potential advantage of using carefully selected background
      sequences is that it may enrich desired signals by normalizing out
      known or even unknown sequence bias. However, a background
      sequence set may not be available or not large enough to for reliable
      estimation of the expected probability of success. A small set of background
      sequences can lead to noisy output, especially for longer <i>k</i>-mers. To avoid a zero estimation of expected probability, a pseudo-count is added to <i>k</i>-mer counts, which can be changed using <code>-pseudo</code>. The default value of the pseudo-count is <code>1</code>.</p>
      <b>Shuffle
        input sequences</b>.
      This is implemented in the option <code>-shuffle n,m</code> (shuffle each sequence <code>n</code> times
      preserving <code>m</code>-residue frequency). <i>k</i>pLogo uses the <a
        href="http://digital.cs.usu.edu/%7Emjiang/ushuffle/">ushuffle</a>
      code to shuffle input sequences while preserving the frequency of
      all
      possible residue combinations of length specified by users. For
      example,
      in many genomic applications, DNA sequences are shuffled while
      preserving di-nucleotide frequencies. Each sequence can be shuffled
      multiple times (default = 100) to enable reliable estimation of the
      background probability of sequences containing the <i>k</i>-mer at a
      particular
      position. However, this will lead to a linear increase in run
      time. In
      addition, for relatively short sequences, it is impossible to
      preserve
      even dinucleotide frequencies. As when using background sequences, a pseudo-count is used to avoid zero estimation of expected probability.<br>
      
      <p>
      <b>Markov
        model from input sequences.</b>
      This is activated by <code>-markov</code>
      specifying the order of the Markov model, which summarizes residue
      frequency and dependence between neighboring residues in input
      sequences. Currently <i>k</i>pLogo supports zero-, first-, and second-order
      Markov
      models. The hypothesized probability of a sequence containing a
      particular <i>k</i>-mer at a particular position is then determined by the Markov model.
      <br>
      <br>
      For example, the first-order Markov model learned from input
      sequences
      includes all mono- and di-residue frequencies, which directly serve
      as
      the hypothesized probabilities for <i>k</i>-mers of length 1 and 2. For
      longer
      <i>k</i>-mers, the hypothesized probability is calculated using the
      property of
      Markov models. For example, the probability of observing the <i>k</i>-mer
      ABCD at a particular position is <br>
      <br>
      <div style="border: 0px solid ; padding: 10px 40px; background:
        rgb(221, 221, 221) none repeat scroll 0% 50%;
        -moz-background-clip: -moz-initial; -moz-background-origin:
        -moz-initial; -moz-background-inline-policy: -moz-initial;
        font-family: Monaco;"><small>p(ABCD) = p(AB) x p(BC)/p(B) x
          p(CD)/p(C) = p(AB) x p(BC) x p(CD) / p(B) / p(C)</small> </div>
      <br>
      Where p(AB), p(BC), p(CD), p(B), and p(C) are the fraction of AB,
      BC,
      and CD di-residues and the fraction of the residue B and C in the
      input
      sequences, respectively.<br>
      <br>
      For zero-order Markov model:<br>
      <br>
      <div style="border: 0px solid ; padding: 10px 40px; background:
        rgb(221, 221, 221) none repeat scroll 0% 50%;
        -moz-background-clip: -moz-initial; -moz-background-origin:
        -moz-initial; -moz-background-inline-policy: -moz-initial;
        font-family: Monaco;"><small>p(ABCD) = p(A) x p(B) x p(C) x p(D)</small>
      </div>
      <br>
      For second-order Markov model:<br>
      <br>
      <div style="border: 0px solid ; padding: 10px 40px; background:
        rgb(221, 221, 221) none repeat scroll 0% 50%;
        -moz-background-clip: -moz-initial; -moz-background-origin:
        -moz-initial; -moz-background-inline-policy: -moz-initial;
        font-family: Monaco;"><small>p(ABCD) = p(ABC) x p(BCD) / p(BC)</small><br>
      </div>
      <br>
      The advantage over shuffling sequences is that it can be much
      faster
      and / or less noisy. However, at this moment Markov models do not
      accurately handle positional shifts and degenerate nucleotides
      (email
      authors for details), and does not support higher-order Markov
      models. Please choose other background models when either
      positional shift or degenerate residues is used. <br>
      <p><b>Markov
        model from background sequences.</b>
      This is activated when both <code>-markov</code> and <code>-bgfile</code> are specified. The only difference from <code>-markov</code> alone is that
      the Markov model is learned from background sequences specified by <code>-bgfile</code>.
      <p><b>Average across all positions (default).
      </b>This is activated by default if none of the above models is
      specified (<code>-bgfile, -shuffle, -markov</code>).
      It is preferred because it is fast, less noisy, and handles both
      positional
      shift and degenerate residues. In this model, the hypothesized
      probability of a sequence with a certain <i>k</i>-mer at a certain
      position is
      the average observed success rate (fraction of input sequences
      containing this <i>k</i>-mer at a given position) averaged over all
      positions.      </p>
      <h3><a class="mozTocH3" name="weighted"></a>Weighted: two-sample Student's <i>t</i>-test</h3>
When turned on by <code>-weighted</code>, <i>k</i>pLogo
      searches for <i>k</i>-mers at certain positions whose presence / absence
      in
      a sequence is associated with higher / lower weights, respectively.
      For
      each <i>k</i>-mer at each position, all input sequences are divided into
      two
      sets: <br>
      <br>
1. positive sequences: sequences containing this <i>k</i>-mer at this
      position,
      allowing position shift and degenerate residues, if specified.<br>
2. negative sequences: all other sequences, i.e., sequences that
      do not contain this <i>k</i>-mer at this position.<br>
      <br>
A <code>one-sided two-sample Student's <i>t</i> test</code> is used to compare the weights of the two sets of sequences. The
      sign
      of the <i>t</i> statistics determines whether it is an enrichment
      (sequences
      with this <i>k</i>-mer at this position have larger weight) or depletion.<br>
      <br>
Note that the input sequences need not to be sorted by weight.<br>
<h3><a class="mozTocH3" name="ranked"></a>Ranked: Mann-Whitney <i>U</i> test</h3>
When
      turned on (by <code>-ranked</code>), <i>k</i>pLogo assumes the input sequences (FASTA
      or
      Tabular) are ranked, and enrichment / depletion of a <i>k</i>-mer at a
      position
      means sequences with this <i>k</i>-mer at this position tend to be ranked
      at
      the top / bottom, respectively. <br>
      <br>
Similar to weighted sequences, ranked sequences are divided into a
      positive set and a negative set based on the presence / absence of
      a <i>k</i>-mer of interest. Instead of Student's t-test, the nonparametric <code>Wilcoxon rank-sum test</code>,
      also called the <code>Mann-Whitney <i>U</i> test</code>,
      is used to determine the significance of positional enrichment /
      depletion of a <i>k</i>-mer. Assuming sample size larger than 8, the
      distribution of the U statistics is approximated by a normal distribution. The resultant z-score and <i>P</i> value from the normal
      approximation is reported in the output. Again the sign of the
      z-score
      determines enrichment or depletion.<br>
      <br>
Note that this model is activated by the option <code>-ranked</code>, which
      will
      ignore <code>-weight</code> if specified. Also note that ranked sequences need to be
      sorted prior to <i>k</i>pLogo analysis.
      <h3><a class="mozTocH3" name="other-parameters"></a>Other parameters<br>
      </h3>
      <p><b> Min count/fraction </b> (<code>-minCount</code>): minimum number or fraction of sequences to have this <i>k</i>-mer to include in output. Default is 5. If a value between 0 and 1 is used, it will be interpreted as the minimum fraction. 
      <br><br>
      <b> Pseudo-count </b> (<code>-pseudo</code>): pseudo-count added to background counts. default = 1. Ignored by Markov models <code>-markov</code>
      <br><br>
      <b> Small-sample correction</b> (<code>-small</code>): used for generating information content logos. See <a href="http://www.sciencedirect.com/science/article/pii/0022283686901658">here</a>.
</p>
      <p>&nbsp; </p>
      <h1><a class="mozTocH1" name="output"></a>Output</h1>
      
		  <p><i>k</i>pLogo outputs both sequence logo plots and tabular data files.</p>
      
            <p style="text-align:center;"><img src="./manual/output.png" width="90%"></p>


	  <p>&nbsp;</p>

      <h2><a class="mozTocH2" name="logo-plots"></a>Logo plots</h2>
      <h3><a class="mozTocH3" name="frequency-logo"></a>Frequency logo</h3>
      <p>In frequency logos, residues are scaled relative to their frequencies at each position, and then stacked on top of each other with the more frequent residue on top of less frequent residue.</p>
      <h3><a class="mozTocH3" name="information-content-logo"></a>Information-content logo</h3>
	<p>Similar to frequency logos, in an information-content logo (conventional sequence logo) residues are scaled by frequency and stacked, and the total height of each position is further scaled relative to the information content, which is defined as </p>
<img src="./infocont.png">

<p>Here N is the number of residues in the alphabet (all possible residues that can occur at a position), and p<small>i</small> is the frequency of residue i at this position. </p> 
      <h3><a class="mozTocH3" name="probability-logo"></a>Probability logo</h3>
<p>Probability logo is first proposed and implemented in <a href="https://plogo.uconn.edu/">pLogo</a> for unweighted sequences, especially protein sequences. <i>k</i>pLogo extends the concept to ranked and weighted sequences. In probability logos, residues are scaled relative to the statistical significance (-log10(<i>P</i> value)) of each residue at each position. Enriched residues stack on the top, whereas depleted residues stack on the bottom. Significant positions  have coordinates colored in red.</p>
      <h3><a class="mozTocH3" name="kmer-logo"></a><i>k</i>-mer logo</h3>
	<p><i>k</i>pLogo generates a new type of logo called <i>k</i>-mer logo to visualize motifs in addition to single residues. Stacked residues at a position represent a single motif (the most significant) starting (or ending) at this position. The total height is scaled relative to the significance of this motif. At each position, two motifs are shown, the most enriched on the top, and the most depleted on the bottom. Both read from top to bottom.</p>
	<h3><a class="mozTocH3" name="p-value-cutoff"></a>Significant positions</h3>
    <p>In probability logo and <i>k</i>-mer logo plots, positions containing residues with <i>Bonferroni</i> corrected <i>P</i> value smaller than a predefined cut-off will be highlighted by <font color=Red>red</font> coordinates. The default cutoff is 0.01 in the webserver. </p>
	<h3><a class="mozTocH3" name="fix"></a>Fixed positions</h3>
	<p>In many applications the input sequences may contain almost invariant residues at certain positions, such as GG in Cas9 guide-RNA target sequences (see example below). The statistical model would not work well for extremely imbalanced samples, or the p-values are so small that it is hard to see signal at other positions. <i>k</i>pLogo allows users to fix positions where a single residue occurs in more than 75% of the sequences. This maximum frequency can be changed using <code>-fix</code> option. <i>P</i> values for fixed positions will be ignored for probability logos, and the height of the fixed residue will be 1.1 times of the max total height of non-fixed positions. The coordinates for fixed positions will be highlighted by black coordinates.</p>
	
      <h3><a class="mozTocH3" name="stackorder"></a>Stacking order</h3>
	<p>In all logo plots other than <i>k</i>-mer logo, multiple residues can stack on top of each other. By default, residues with larger weights will be placed on top of residues with smaller weights. This can be reversed with <code>-stack_order -1</code>. To ignore weight information and use order specified by alphabet, use <code>-stack_order 0</code>.  </p>
	<h3><a class="mozTocH3" name="scaling"></a>Scaling</h3>
    <p>By default residues or <i>k</i>-mers are scaled by -log10 transformed <i>P</i> values. To highlight more the significant motifs, one can plot Bonferroni corrected <i>P</i> values. One can also plot test statistics. In cases where <i>P</i> values are too small to be represented in the computer system (P < 1E-308 ), <i>k</i>pLogo will automatically switch to test statistics.      </p>
    <p style="text-align:center;"><img src="./manual/scaling.png" width="40%"></p>

	
     
      <h3><a class="mozTocH3" name="kmer-position"></a><i>k</i>-mer coordinates</h3>
	  
	  By default, <i>k</i>pLogo assigns each <i>k</i>-mer to the position of its starting letter (i.e., most 5' or most N-terminal)  and compares the significance of its enrichment to that of all other <i>k</i>-mers assigned to the same position. <i>k</i>pLogo also allows the option of assigning each <i>k</i>-mer to the position of its end letter. Comparing these two schemes using pri-miRNA sequences shows that known motifs sharing a start or end position with a stronger motif can be masked out in one scheme or the other but that the strongest motif among all overlapping ones was identified in both schemes. 
	  
	  <h3><a class="mozTocH3" name="position1"></a>Position 1</h3>
    <p>  In many cases it makes more sense to call some position other than the first residue as position 1. If this is the case the positions before it will be numbered as -1, -2, etc (no 0).</p>
    <p style="text-align:center;"><img src="./manual/position1.png" width="80%"></p>
    
     
      <h2><a class="mozTocH2" name="tabular-data-file"></a>Tabular data
        file</h2>
      <p>Regardless the type of input (unweighted/ranked/weighted), the
        first six columns of the output data file are always the same:</p>
      <div style="border: 0px solid ; padding: 10px 40px; background:
        rgb(221, 221, 221) none repeat scroll 0% 50%;
        -moz-background-clip: -moz-initial; -moz-background-origin:
        -moz-initial; -moz-background-inline-policy: -moz-initial;
        font-family: Monaco;"><small>
          <ol>
            <li><i>k</i>-mer</li>
            <li>position: first nucleotide of the <i>k</i>-mer in the sequence.</li>
            <li>shift</li>
            <li>test statistics (positive for enrichment and negative for
              depletion)</li>
            <li><i>P</i> value (-log10)</li>
            <li>Bonferroni corrected <i>P</i> value (-log10)</li>
          </ol>
        </small> </div>
      <p>Output from ranked inputs has no other columns other than the
        six above. Note the statistics and <i>P</i> value in column 4 and 5 are
        based on normal approximation. See: </p>
      <div style="border: 0px solid ; padding: 10px 40px; background:
        rgb(221, 221, 221) none repeat scroll 0% 50%;
        -moz-background-clip: -moz-initial; -moz-background-origin:
        -moz-initial; -moz-background-inline-policy: -moz-initial;
        font-family: Monaco;"><small>
          Carine A. Bellera et al, Normal Approximations to the
          Distributions of the Wilcoxon Statistics: Accurate to What N?
          Graphical Insights, <strong>Journal of
            Statistics Education</strong>, Volume 18, Number 2, (2010) <a
href="http://www.amstat.org/publications/jse/v18n2/bellera.pdf">Download PDF</a>
        </small></div>
      <p>Unweighted inputs will generate four additional columns in the
        output:</p>
      <div style="border: 0px solid ; padding: 10px 40px; background:
        rgb(221, 221, 221) none repeat scroll 0% 50%;
        -moz-background-clip: -moz-initial; -moz-background-origin:
        -moz-initial; -moz-background-inline-policy: -moz-initial;
        font-family: Monaco;"><small>
          <ol start="7">
            <li>The observed fraction of input sequences with this <i>k</i>-mer at
              this position allowing this amount of shift</li>
            <li>The expected fraction of sequences with this <i>k</i>-mer
              at this position allowing this amount of shift</li>
            <li>ratio: column 7 divided by column 8</li>
            <li>column 7 divided by the average of other positions </li>
          </ol>
        </small> </div>
      <p></p>
      Weighted inputs will generate 6 more columns:
      <p></p>
      <div style="border: 0px solid ; padding: 10px 40px; background:
        rgb(221, 221, 221) none repeat scroll 0% 50%;
        -moz-background-clip: -moz-initial; -moz-background-origin:
        -moz-initial; -moz-background-inline-policy: -moz-initial;
        font-family: Monaco;"><small>
          <ol start="7">
            <li>sample size for sample 1, i.e., the number of sequences
              that have this <i>k</i>-mer at this position when allowing this amount
              of shift. </li>
            <li>mean of weight for sample 1 sequences.</li>
            <li>standard deviation of weight for sample 1 sequences.</li>
            <li>sample size for sample 2, i.e., the number of sequences
              that do not
              have this <i>k</i>-mer at this position when allowing this amount
              of shift. </li>
            <li>mean of weight for sample 2 sequences.</li>
            <li>standard deviation of weight for sample 2 sequences.</li>
          </ol>
        </small> </div>
		
      <h2><a class="mozTocH2" name="run-time"></a>Run time and input limit</h2>
	  
	  <p>
		  
		  For typical DNA/RNA analysis, the job should finish within a minute. Run time increases dramatically with increased <i>k</i>-mer length, and even more with protein sequences. The program imposes a limit of 10 million tests per job, which equals the number of <i>k</i>-mers multiplied by the number of positions (length of each sequence). On the webserver, the largest number of <i>k</i>-mers to be tested for protein is 20^1+20^2+20^3=8420 for <i>k</i>-mer of length (1,2,3), which limits the maximum length of each input sequence to 1188 (10,000,000 / 8420), a number large enough for most applications. This limit becomes much larger if users choose to use shorter <i>k</i>-mers. Similarly, for DNA/RNA sequence analysis, the limit on input sequence length is 1832 for <i>k</i>-mer length of (1,2,3,4,5,6) [=10,000,000/(4^1+4^2+4^3+4^4+4^5+4^6)] without considering degenerate letters. The limit will vary dramatically if users choose to use different degenerate letters. If the limit is reached, the webserver reports an error for too many tests, and users can go back and choose shorter <i>k</i>-mers and / or omit degenerate letters. 
		  
		  <br><br>
		  
		  Here are a few examples:
		  
		  <br><br>
		  42,481 protein sequences, each 40 amino acids long: It took 8 seconds, 1.5 minutes, and 34 minutes to run with a <i>k</i>-mer length up to 1, 2, and 3, respectively. 
		  <br><br>
		  186 RNA sequences, each 50 nt long: It took 6 seconds to run with a <i>k</i>-mer length up to 4, allowing gaps (i.e. the degenerate letter N), but 14 minutes if allowing all degenerate letters (ACGTRYMKWSBDHVN).
		  
	  </p>
      
      <h2><a class="mozTocH2" name="galaxy"></a>Galaxy Tool Shed</h2>
	  
	  <p>
		  
	  	<i>k</i>pLogo has been wrapped as a Galaxy tool and is available in Galaxy Tool Shed (<a href="https://toolshed.g2.bx.psu.edu/view/xuebing/kplogo">click here</a>). To run <i>k</i>pLogo in a local instance of Galaxy, one still needs to compile and install <i>k</i>pLogo locally (see next section), and add the executable <i>k</i>pLogo to the path, such that one can directly call <i>k</i>pLogo by typing "kpLogo" in a terminal window.  
		
	  </p>
		
      <h2><a class="mozTocH2" name="install-kpLogo-locally"></a>Install <i>k</i>pLogo
        locally</h2>
      <p><i>k</i>pLogo is an open-source project and users can download the source
        code, compile and install locally to enable all functionality.</p>
      <h3><a class="mozTocH3" name="requirements"></a>Requirements</h3>
      <i>k</i>pLogo has been successfully tested on Linux and Mac.<br>
      <br>
      <span style="font-weight: bold;">To compile: GCC</span><br>
      <i>k</i>pLogo is written in C++ and reuses some libraries from <a
        href="http://www.boost.org/">Boost</a> version 1.57.0. To
      compile <i>k</i>pLogo
      locally, one should have a recent version of <a
        href="https://gcc.gnu.org/">GCC</a> that is compatible with the
      g++
      option <code>-std=c++11</code>.
      <br>
      <br>
      <span style="font-weight: bold;">To generate PNG and PDF files:
        ImageMagick &amp; Ghostscript</span><br>
      Logos are generated in postscript (ps) files and then converted to
      PNG
      using the program <span style="color: rgb(51, 51, 255);">convert</span>
      in the package <span style="color: rgb(51, 51, 255);">ImageMagick</span>.
      In addition, <span style="color: rgb(51, 51, 255);">Ghostscript</span>
      is used to convert postscript files to PDF files.<br>
      <br>
      <h3><a class="mozTocH3" name="download"></a>Download</h3>
      The <i>k</i>pLogo source code (including web code) is hosted on GitHub: <a
        href="https://github.com/xuebingwu/kpLogo">https://github.com/xuebingwu/kpLogo</a><br>
      <br>
      To clone the latest (may not be stable) version:<br>
      <br>
      <div style="border: 0px solid ; padding: 10px 40px; background:
        rgb(221, 221, 221) none repeat scroll 0% 50%;
        -moz-background-clip: -moz-initial; -moz-background-origin:
        -moz-initial; -moz-background-inline-policy: -moz-initial;
        font-family: Monaco;"><small>git
            clone
            https://github.com/xuebingwu/kpLogo.git</small></div>
      <br>
      <br>
      To download stable releases of <i>k</i>pLogo: <a
        href="https://github.com/xuebingwu/kpLogo/releases/">https://github.com/xuebingwu/kpLogo/releases/
      </a><br>
      <br>
      <h3><a class="mozTocH3" name="installation"></a>Installation</h3>
      <br>
      Once finished downloading, unzip if necessary, enter the subfolder
      <code>src</code>,
      then type <code>make</code>: <br>
      <br>
      <div style="border: 0px solid ; padding: 10px 40px; background:
        rgb(221, 221, 221) none repeat scroll 0% 50%;
        -moz-background-clip: -moz-initial; -moz-background-origin:
        -moz-initial; -moz-background-inline-policy: -moz-initial;"><small><span
              style="font-family: Monaco;">cd src</span><br
              style="font-family: Monaco;">
            <span style="font-family: Monaco;">make</span></small>
          <br>
      </div>
      <br>
      The compiled programs are in the folder <code>bin</code>.<br>
      To recompile, type <code>make clean</code> then <code>make</code>.<br>
      <br>
      
 

      <h2><a class="mozTocH2" name="command-line-options"></a>Command-Line Options</h2>

      <p>Click <a href="command-line.usage">here</a> to view the command-line options in plain text</p>
	  <p>An enhanced html version is coming soon.</p>
	  <p>&nbsp;</p>
      
        <h1><a class="mozTocH1" name="examples"></a>Examples</h1>
		Here are two examples discussed in the <i>k</i>pLogo paper: 
		
        <div style="border: 0px solid ; padding: 10px 40px; background:
          rgb(222, 222, 222) none repeat scroll 0% 50%;
          -moz-background-clip: -moz-initial; -moz-background-origin:
          -moz-initial; -moz-background-inline-policy: -moz-initial;
          font-family: Monaco;">
		  <p>
	      Xuebing Wu and David Bartel (2017) <i>k</i>pLogo: positional <i>k</i>-mer analysis reveals hidden specificity in biological sequences, <em><strong>Nucleic Acids Research</strong></em>, in press.</p>
	   </div>
	   
		<h2><a class="mozTocH2" name="example1"></a>1. Unweighted input: sequence features defining human microRNA genes </h2>
		
        <h4><a class="mozTocH4" name="example1-input"></a>Data set </h4>
						
		50 nt centered on the 3' end cleavage site of human microRNA hairpin <a href="./examples/miRNA-3p-50nt/premir50p3.fa" target="_blank"> [Download] </a>  <br>

        <h4><a class="mozTocH4" name="example1-setting"></a>Parameter setting </h4>

		<p>Web: no need to change default settings. If you want to re-number the coordinates such that the two nucleotides flanking the cleavage site are numbered as -1 and 1, respectively, set 'Position 1' to be 26 in the 'Output' option groups. </p>
		
		<p>Full command line:</p>
         <div style="border: 0px solid ; padding: 10px 40px; background:
           rgb(222, 222, 222) none repeat scroll 0% 50%;
           -moz-background-clip: -moz-initial; -moz-background-origin:
           -moz-initial; -moz-background-inline-policy: -moz-initial;
           font-family: Monaco;">
 		  <p>
		   kpLogo premir50p3.fa -gapped -startPos 26 -pc 0.01
 	      </p>
 	   </div>
		

        <h4><a class="mozTocH4" name="example1-result"></a>Output </h4>
		
		<p> <i>k</i>pLogo outputs four types of logos:</p>

        <div style="border: 0px solid ; padding: 10px 40px; background:
          rgb(222, 222, 222) none repeat scroll 0% 50%;
          -moz-background-clip: -moz-initial; -moz-background-origin:
          -moz-initial; -moz-background-inline-policy: -moz-initial;
          font-family: Monaco;"><small>

<table border-spacing="0" border-collapse="collapse" display="block" cellspacing="0" cellpadding="0" border="0">
	<tbody>
		<tr>
			<td><strong>Frequency Logo</strong></td>
			<td><a href="./examples/miRNA-3p-50nt/kpLogo.freq.png">  <img src="./examples/miRNA-3p-50nt/kpLogo.freq.png" style="display:block" width="100%"> </a>
			</td>
		</tr>
		
		<tr>
			<td><strong>Information Content Logo</strong></td>
			<td><a href="./examples/miRNA-3p-50nt/kpLogo.info.png">  <img src="./examples/miRNA-3p-50nt/kpLogo.info.png" style="display:block" width="100%"> </a>
			</td>
		</tr>
		
		<tr>
			<td><strong>Probability Logo</strong></td>
			<td><a href="./examples/miRNA-3p-50nt/kpLogo.png">  <img src="./examples/miRNA-3p-50nt/kpLogo.png" style="display:block" width="100%"> </a>
			</td>
		</tr>
		
		<tr>
			<td><strong><i>k</i>-mer Logo</strong></td>
			<td><a href="./examples/miRNA-3p-50nt/kpLogo.most.significant.each.position.png">  <img src="./examples/miRNA-3p-50nt/kpLogo.most.significant.each.position.png" style="display:block" width="100%"> </a>
			</td>
		</tr>
	</tbody>
	</table>
		
		    </small> </div>

		<p> </p>

        <h2><a class="mozTocH2" name="example2"></a>Weighted/Ranked: Sequence features for CRISPR/Cas9 guide RNA efficiency </h2>
		
		        <h4><a class="mozTocH4" name="example1-input"></a>Data set </h4>
						
				30-nucleotide sequences around target regions of 1841 guide RNAs and associated cutting efficiency <a href="./examples/gRNA-efficiency/gRNA.txt" target="_blank"> [Download] </a>  <br>

		        <h4><a class="mozTocH4" name="example1-setting"></a>Parameter setting </h4>

				<p>Web: select 'Weighted' or 'Ranked' in the 'Input' option groups. They give similar results. If you want to re-number the coordinates such that the first nucleotide of the guide-RNA match is numbered as 1, set 'Position 1' to be 5 in the 'Output' option groups. </p>
		
				<p>Full command line:</p>
		         <div style="border: 0px solid ; padding: 10px 40px; background:
		           rgb(222, 222, 222) none repeat scroll 0% 50%;
		           -moz-background-clip: -moz-initial; -moz-background-origin:
		           -moz-initial; -moz-background-inline-policy: -moz-initial;
		           font-family: Monaco;">
		 		  <p>
				   kpLogo gRNA.txt -weighted -gapped -startPos 5 -pc 0.01 <br>
				   kpLogo gRNA.txt -ranked -gapped -startPos 5 -pc 0.01
		 	      </p>
		 	   </div>
		

		        <h4><a class="mozTocH4" name="example1-result"></a>Output </h4>
		
				<p> <i>k</i>pLogo outputs four types of logos:</p>

		        <div style="border: 0px solid ; padding: 10px 40px; background:
		          rgb(222, 222, 222) none repeat scroll 0% 50%;
		          -moz-background-clip: -moz-initial; -moz-background-origin:
		          -moz-initial; -moz-background-inline-policy: -moz-initial;
		          font-family: Monaco;"><small>

		<table border-spacing="0" border-collapse="collapse" display="block" cellspacing="0" cellpadding="0" border="0">
			<tbody>
				<tr>
					<td><strong>Frequency Logo</strong></td>
					<td><a href="./examples/gRNA-efficiency/kpLogo.freq.png">  <img src="./examples/gRNA-efficiency/kpLogo.freq.png" style="display:block" width="100%"> </a>
					</td>
				</tr>
		
				<tr>
					<td><strong>Information Content Logo</strong></td>
					<td><a href="./examples/gRNA-efficiency/kpLogo.info.png">  <img src="./examples/gRNA-efficiency/kpLogo.info.png" style="display:block" width="100%"> </a>
					</td>
				</tr>
		
				<tr>
					<td><strong>Probability Logo</strong></td>
					<td><a href="./examples/gRNA-efficiency/kpLogo.png">  <img src="./examples/gRNA-efficiency/kpLogo.png" style="display:block" width="100%"> </a>
					</td>
				</tr>
		
				<tr>
					<td><strong><i>k</i>-mer Logo</strong></td>
					<td><a href="./examples/gRNA-efficiency/kpLogo.most.significant.each.position.png">  <img src="./examples/gRNA-efficiency/kpLogo.most.significant.each.position.png" style="display:block" width="100%"> </a>
					</td>
				</tr>
			</tbody>
			</table>
		
				    </small> </div>

				
        <p><br>
        </p>
      <h4><a class="mozTocH4" name="example1-setting"></a>Prediction</h4>
        <p>The commandline version of <em>k</em>pLogo supports scoring of a list of new sequences using positional <em>k</em>-mer motifs learned from a previous run. Briefly, for a given sequence to be scored, the score is the sum of log10(p-value) of all positional <em>k</em>-mers present in the sequence. </p>
        <p>As a toy example, here we will score the guide RNA sequences using the positional motifs we learned from themselves: </p>
        <p>Full command line:</p>
      <div style="border: 0px solid ; padding: 10px 40px; background:
		           rgb(222, 222, 222) none repeat scroll 0% 50%;
		           -moz-background-clip: -moz-initial; -moz-background-origin:
		           -moz-initial; -moz-background-inline-policy: -moz-initial;
		           font-family: Monaco;">
        <p># training <br>
        kpLogo gRNA.txt -weighted -o model_training</p>
        <p># prediction<br>
          kpLogo gRNA.txt -predict model_training</p>
        <p># the output file model_training.score contains the score for each sequence</p>
        </div>
        <p> </p>
		

      <h1><a class="mozTocH1" name="copyright"></a>Copyright</h1>
      <p>Copyright (c) 2017 Xuebing Wu. </p>
      <p><i>k</i>pLogo reuses the following library/code:<br>
        - <a href="http://www.boost.org/">Boost</a> library (version
        1.57.0)<br>
        - The <a href="http://digital.cs.usu.edu/%7Emjiang/ushuffle/">ushuffle</a>
        code written by Minghui Jiang <br>
      </p>
      <div style="border: 0px solid ; padding: 10px 40px; background:
        rgb(221, 221, 221) none repeat scroll 0% 50%;
        -moz-background-clip: -moz-initial; -moz-background-origin:
        -moz-initial; -moz-background-inline-policy: -moz-initial;
        font-family: Monaco;"><a>
          <p><b>License</b><br>
            <br>
            <small>
              Permission is hereby granted, free of charge, to any
              person obtaining a copy of this software and associated
              documentation files (the "Software"), to deal in the
              Software without restriction, including without limitation
              the rights to use, copy, modify, merge, publish,
              distribute, sublicense, and/or sell copies of the
              Software, and to permit persons to whom the Software is
              furnished to do so, subject to the following conditions:</small></p>
          <small>
            <p>The above copyright notice and this permission notice
              shall be included in all copies or substantial portions of
              the Software.</p>
          </small>
          <p><small>
              THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY
              KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
              WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
              PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS
              OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
              OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
              OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
              SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.</small></p>
        </a></div>
      <h1><a class="mozTocH1" name="author"></a>Author</h1>
      <p><i>k</i>pLogo is developed and maintained by Xuebing Wu in the Bartel lab
        at Whitehead Institute.</p>
      <h1><a class="mozTocH1" name="acknowlegements"></a>Acknowledgements</h1>
      <p>I would like to thank Whitehead Institute IT system team (especially Paul McCabe, Andy Nutter-Upham, Craig Andrew, Alexan Mardigian) for help with the server setup. </p>
      <h1><a class="mozTocH1" name="reference"></a>Reference</h1>
      <p>Xuebing Wu and David Bartel (2017) <i>k</i>pLogo: positional <i>k</i>-mer analysis reveals hidden specificity in biological sequences, <em><strong>Nucleic Acids Research</strong></em>, in press.</p>
    </div>
  </body>
</html>
