---
# WARNING! This file is used in docker container only. The container uses binary r-cran packages of ubuntu 22.04. These packages are old and not compatible with current R versions. The draw_figures.Rmd is run in native.
title: "Rearrangement report"
output: html_document
fig_width: 1920px
params:
  algfile: ""
---

```{r load R library, include = FALSE}
library(tidyverse)
library(ggseqlogo)
library(ggforce)
library(waffle)
library(patchwork)
library(reticulate)
library(this.path)
```

```{python import python package, include = FALSE}
import numpy
import scipy
import os
```

```{r set parameters for R, include = FALSE}
algfile <- params$algfile
tablefile <- sub(".alg.", ".table.", algfile)
fields <- strsplit(algfile, ".", fixed = TRUE)[[1]]
ref1len <- strtoi(last(fields))
cut1 <- strtoi(fields[length(fields) - 2])
cut2 <- strtoi(fields[length(fields) - 1])
lineNum <- strtoi(system(sprintf("wc -l <%s", algfile), intern = TRUE))
```

```{r load data, include = FALSE}
counts <- scan(file = pipe(sprintf("sed -n '1~3p' %s | cut -f2", algfile)), what = integer())
reflines <- readLines(con = pipe(sprintf("sed -nr '2~3p' %s", algfile)))
querylines <- readLines(con = pipe(sprintf("sed -nr '3~3p' %s", algfile)))
ref1 <- reflines[1] |> gsub(pattern = "-", replacement = "", fixed = TRUE) |> substring(1, ref1len) |> toupper()
ref2 <- reflines[1] |> gsub(pattern = "-", replacement = "", fixed = TRUE) |> substring(ref1len + 1) |> toupper()
ref1vec <- str_split(ref1, "")[[1]]
ref2vec <- str_split(ref2, "")[[1]]
```

```{python assign data, include = FALSE}
counts = r.counts
reflines = r.reflines
querylines = r.querylines
```

```{python calculate base frequency, include = FALSE}
basefreq, alphabet = numpy.zeros((5, len(r.ref1) + len(r.ref2)), dtype = numpy.int64), numpy.reshape(numpy.frombuffer(b"ACGT-", dtype = numpy.int8), (5, 1))
for i in range(len(counts)):
    querybases = numpy.frombuffer(querylines[i].encode(), dtype = numpy.int8)[numpy.frombuffer(reflines[i].encode(), dtype = numpy.int8) != ord('-')]
    basefreq += (querybases == alphabet) * counts[i]
_ = 0 # handle block bug
```

```{python calculate insertcount, include = FALSE}
insertcount = numpy.zeros(len(r.ref1) + len(r.ref2) + 1, dtype = numpy.int64)
for i in range(len(reflines)):
    mask = numpy.frombuffer(reflines[i].encode(), dtype = numpy.int8) != ord('-')
    insertcount += numpy.bincount(numpy.cumsum(mask)[mask == False], minlength=len(insertcount)) * counts[i]
_ = 0 # handle block bug
```

```{python calculate MSDfreq, include = FALSE}
MSDfreq = numpy.zeros((3, len(r.ref1) + len(r.ref2)), dtype = numpy.int64)
for i in range(len(counts)):
    refbases = numpy.frombuffer(reflines[i].upper().encode(), dtype = numpy.int8)
    mask = (refbases != ord('-'))
    querybases = numpy.frombuffer(querylines[i].encode(), dtype = numpy.int8)[mask]
    refbases = refbases[mask]
    MSDfreq[0, :] += (refbases == querybases) * counts[i]
    MSDfreq[1, :] += numpy.logical_and(querybases != ord('-'), refbases != querybases) * counts[i]
    MSDfreq[2, :] += (querybases == ord('-')) * counts[i]
_ = 0 # handle block bug
```

```{python calculate TRANfreq, include = FALSE}
poly = scipy.interpolate.lagrange(numpy.array([ord(cc) for cc in ["-", "A", "C", "G", "T"]]), numpy.arange(5))
TRANfreq, refbases, querybases, repcounts = numpy.zeros((5, 5), dtype = numpy.int64), [], [], []
for i in range(len(counts)):
    refbases.append(numpy.round(poly(numpy.frombuffer(reflines[i].upper().encode(), dtype = numpy.int8))).astype(numpy.int8))
    querybases.append(numpy.round(poly(numpy.frombuffer(querylines[i].encode(), dtype = numpy.int8))).astype(numpy.int8))
    repcounts.append(len(refbases[-1]))
TRANfreq = scipy.sparse.csc_array((numpy.repeat(counts, repcounts), (numpy.concatenate(refbases), numpy.concatenate(querybases))), shape = (5, 5)).toarray()
_ = 0 # handle block bug
```

```{r barplot of positional base for ref1, fig.cap = "barplot of positional base for ref1", echo = FALSE, message = FALSE, warning = FALSE}
tibble(count = c(py$basefreq), pos = rep(seq(ncol(py$basefreq)) - 0.5, each = nrow(py$basefreq)), base = rep(c("A", "C", "G", "T", "-"), times = ncol(py$basefreq))) |>
  filter(pos <= ref1len) |>
  mutate(base = factor(base, levels = c("-", "A", "C", "G", "T"))) |>
  mutate(rel1pos = pos - cut1) |>
  ggplot(aes(rel1pos, count)) +
  geom_col(aes(fill = base)) +
  geom_step(aes(pos - cut1, count, color = "black"), data = tibble(pos = 0:(nchar(ref1)+nchar(ref2)), count = py$insertcount) |> filter(pos >0, pos < ref1len), direction = "mid") +
  scale_x_continuous(name = "position relative to cut1", expand=c(0,0)) +
  scale_y_continuous(expand=c(0,0)) +
  scale_color_identity(name = NULL, guide = guide_legend(), labels = "insertion") ->
  ggfig
print(ggfig)
```

```{r barplot of positional base for ref2, fig.cap = "barplot of positional base for ref2", echo = FALSE, message = FALSE, warning = FALSE}
tibble(count = c(py$basefreq), pos = rep(seq(ncol(py$basefreq)) - 0.5, each = nrow(py$basefreq)), base = rep(c("A", "C", "G", "T", "-"), times = ncol(py$basefreq))) |>
  filter(pos > ref1len) |>
  mutate(base = factor(base, levels = c("-", "A", "C", "G", "T"))) |>
  mutate(rel2pos = pos - ref1len - cut2) |>
  ggplot(aes(rel2pos, count)) +
  geom_col(aes(fill = base)) +
  geom_step(aes(pos - ref1len - cut2, count, color = "black"), data = tibble(pos = 0:(nchar(ref1)+nchar(ref2)), count = py$insertcount) |> filter(pos > ref1len, pos < nchar(ref1)+nchar(ref2)), direction = "mid") +
  scale_x_continuous(name = "position relative to cut2", expand=c(0,0)) +
  scale_y_continuous(expand=c(0,0)) +
  scale_color_identity(name = NULL, guide = guide_legend(), labels = "insertion") ->
  ggfig
print(ggfig)
```

```{r barplot of positional indel for ref1, fig.cap = "barplot of positional indel for ref1", echo = FALSE, message = FALSE, warning = FALSE}
tibble(count = c(py$MSDfreq), pos = rep(seq(ncol(py$MSDfreq)) - 0.5, each = nrow(py$MSDfreq)), type = rep(c("match", "SNP", "delete"), times = ncol(py$MSDfreq))) |>
  filter(pos <= ref1len) |>
  mutate(type = factor(type, levels = c("delete", "SNP", "match"))) |>
  mutate(rel1pos = pos - cut1) |>
  ggplot(aes(rel1pos, count)) +
  geom_col(aes(fill = type)) +
  geom_step(aes(pos - cut1, count), data = tibble(pos = 0:(nchar(ref1)+nchar(ref2)), count = py$insertcount) |> filter(pos >0, pos < ref1len), color = "black", direction = "mid") +
  scale_x_continuous(name = "position relative to cut1", expand=c(0,0)) +
  scale_y_continuous(expand=c(0,0)) +
  scale_color_identity(name = NULL, guide = guide_legend(), labels = "insertion") ->
  ggfig
print(ggfig)
```

```{r barplot of positional indel for ref2, fig.cap = "barplot of positional indel for ref2", echo = FALSE, message = FALSE, warning = FALSE}
tibble(count = c(py$MSDfreq), pos = rep(seq(ncol(py$MSDfreq)) - 0.5, each = nrow(py$MSDfreq)), type = rep(c("match", "SNP", "delete"), times = ncol(py$MSDfreq))) |>
  filter(pos > ref1len) |>
  mutate(type = factor(type, levels = c("delete", "SNP", "match"))) |>
  mutate(rel2pos = pos - ref1len - cut2) |>
  ggplot(aes(rel2pos, count)) +
  geom_col(aes(fill = type)) +
  geom_step(aes(pos - ref1len - cut2, count), data = tibble(pos = 0:(nchar(ref1)+nchar(ref2)), count = py$insertcount) |> filter(pos > ref1len, pos < nchar(ref1)+nchar(ref2)), color = "black", direction = "mid") +
  scale_x_continuous(name = "position relative to cut2", expand=c(0,0)) +
  scale_y_continuous(expand=c(0,0)) +
  scale_color_identity(name = NULL, guide = guide_legend(), labels = "insertion") ->
  ggfig
print(ggfig)
```

```{r base count of dangle, fig.cap = "base count of dangle", echo = FALSE, message = FALSE, warning = FALSE}
tibble(count = py$insertcount[c(1, ref1len + 1, nchar(ref1) + nchar(ref2) + 1)], part = c("updangle", "randomins", "downdangle")) |>
  mutate(part = factor(part, levels = c("updangle", "randomins", "downdangle"))) |>
  ggplot(aes(part, count)) +
  geom_col() +
  scale_y_continuous(expand=c(0,0)) ->
  ggfig
print(ggfig)
```

```{r base substitution, fig.cap = "base substitution", echo = FALSE, message = FALSE, warning = FALSE}
trans <- paste(rep(c("-", "A", "C", "G", "T"), times = 5), rep(c("-", "A", "C", "G", "T"), each = 5), sep = ">")
tibble(count = c(py$TRANfreq), trans = factor(trans, levels = trans)) |>
  ggplot(aes(trans, count)) +
  geom_col() +
  scale_y_continuous(expand=c(0,0)) ->
  ggfig
print(ggfig)
```

```{r show micro-homology, fig.cap = "show micro-homology", fig.asp = 1, echo = FALSE, message = FALSE, warning = FALSE}
microhomo <- matrix(as.integer(rep(ref1vec, time = length(ref2vec)) == rep(ref2vec, each = length(ref1vec))), nrow = length(ref1vec))
for (i in seq_len(nrow(microhomo))) {
  for (j in seq_len(ncol(microhomo))) {
    if (i > 1 && j > 1)
    {
      if (microhomo[i, j] > 0) {
        microhomo[i, j] <- microhomo[i - 1, j - 1] + microhomo[i, j]
      }
    }
  }
}
rc <- which(microhomo > 3, arr.ind = T)
pos12 <- rc[rep(seq_len(nrow(rc)), time = microhomo[rc]),] -
  rep(unlist(sapply(microhomo[rc], function(i) seq(i - 1, 0, by = -1))), time = 2)
tibble(pos1 = pos12[, 1], pos2 = pos12[, 2]) |>
  ggplot(aes(pos1, pos2)) +
  geom_point() +
  scale_x_continuous(expand=c(0,0)) +
  scale_y_continuous(expand=c(0,0)) ->
  ggfig
print(ggfig)
```

```{r load and preprocess information table for alignments, include = FALSE}
indel_tsv <- read_tsv(tablefile, col_types = "iiiciiiiciiiiciic", na = "NA")
indel_tsv |>
  mutate(uplen = query_start1, randlen = query_start2 - query_end1, downlen = nchar(downdangle)) |>
  select(count, score, ref_end1, query_end1, ref_start2, query_start2, downdangle, cut1, cut2, uplen, randlen, downlen) |>
  mutate(templated = (ref_end1 > cut1 | ref_start2 < cut2)) |>
  mutate(deletion = (ref_end1 < cut1 | ref_start2 > cut2)) |>
  mutate(random = (query_end1 < query_start2)) |>
  mutate(insertion = (templated | random)) |>
  mutate(indel_type = factor(ifelse(insertion & !deletion, "insertion",
                             ifelse(deletion & !insertion, "deletion",
                             ifelse(insertion & deletion, "indel", "WT"
  ))), levels = c("WT", "deletion", "insertion", "indel"))) |>
  mutate(indel_type_ex = factor(ifelse(templated & deletion & random, "full",
                             ifelse(templated & deletion & !random, "tempdel",
                             ifelse(templated & !deletion & random, "temprand",
                             ifelse(templated & !deletion & !random, "templated",
                             ifelse(!templated & deletion & random, "randdel",
                             ifelse(!templated & deletion & !random, "deletion",
                             ifelse(!templated & !deletion & random, "random", "WT"
  ))))))), levels = c("WT", "deletion", "templated", "random", "temprand", "tempdel", "randdel", "full"))) ->
  indel_tsv
```

```{r indel type count, fig.cap = "indel type count", echo = FALSE, message = FALSE, warning = FALSE}
indel_tsv |>
  group_by(indel_type) |>
  summarise(count = sum(count)) |>
  ungroup() |>
  mutate(percent = count / sum(count), perlabel = scales::percent(percent, accuracy = 0.01)) |>
  mutate(type_count = sprintf("%s: %d", indel_type, count)) |>
  mutate(type_count = factor(type_count, levels = c(type_count[startsWith(type_count, "WT")], type_count[startsWith(type_count, "deletion")], type_count[startsWith(type_count, "insertion")], type_count[startsWith(type_count, "indel")]))) |>
  ggplot(aes(1, percent, fill = type_count, weight = count)) +
  geom_col() +
  geom_text(aes(label = perlabel), position = position_stack(vjust = 0.5), size = 1.5) +
  scale_x_discrete(name = NULL, breaks = NULL) +
  scale_y_continuous(breaks = c(0, 0.25, 0.5, 0.75), labels = scales::percent) +
  scale_fill_manual(values = RColorBrewer::brewer.pal(8, "Set1")[c(1, 2, 3, 6)]) +
  coord_polar(theta = "y") +
  theme(text = element_text(size = 5)) ->
  ggfig1

indel_tsv |>
  group_by(indel_type_ex) |>
  summarise(count = sum(count)) |>
  ungroup() |>
  mutate(percent = count / sum(count), perlabel = scales::percent(percent, accuracy = 0.01)) |>
  mutate(type_count_ex = sprintf("%s: %d", indel_type_ex, count)) |>
  mutate(type_count_ex = factor(type_count_ex, levels = c(type_count_ex[startsWith(type_count_ex, "WT")], type_count_ex[startsWith(type_count_ex, "deletion")], type_count_ex[startsWith(type_count_ex, "templated")], type_count_ex[startsWith(type_count_ex, "random")], type_count_ex[startsWith(type_count_ex, "temprand")], type_count_ex[startsWith(type_count_ex, "tempdel")], type_count_ex[startsWith(type_count_ex, "randdel")], type_count_ex[startsWith(type_count_ex, "full")]))) |>
  ggplot(aes(1, percent, fill = type_count_ex, weight = count)) +
  geom_col() +
  geom_text(aes(label = perlabel), position = position_stack(vjust = 0.5), size = 1.5) +
  scale_x_discrete(name = NULL, breaks = NULL) +
  scale_y_continuous(breaks = c(0, 0.25, 0.5, 0.75), labels = scales::percent) +
  scale_fill_manual(values = RColorBrewer::brewer.pal(8, "Set1")) +
  coord_polar(theta = "y") +
  theme(text = element_text(size = 5)) ->
  ggfig2

indel_tsv |>
  group_by(indel_type) |>
  summarise(count = sum(count)) |>
  ungroup() |>
  ggplot(aes(fill = indel_type, values = count)) +
  geom_waffle(n_rows = 10, color = "white", make_proportional = TRUE) +
  scale_fill_manual(limits = c("WT", "deletion", "insertion", "indel"), values = RColorBrewer::brewer.pal(8, "Set1")[c(1, 2, 3, 6)]) +
  coord_equal() +
  theme(text = element_text(size = 5)) ->
  ggfig3

indel_tsv |>
  group_by(indel_type_ex) |>
  summarise(count = sum(count)) |>
  ungroup() |>
  ggplot(aes(fill = indel_type_ex, values = count)) +
  geom_waffle(n_rows = 10, color = "white", make_proportional = TRUE) +
  scale_fill_manual(limits = c("WT", "deletion", "templated", "random", "temprand", "tempdel", "randdel", "full"), values = RColorBrewer::brewer.pal(8, "Set1")) +
  coord_equal() +
  theme(text = element_text(size = 5)) ->
  ggfig4

ggfig <- (ggfig1 | ggfig2) / (ggfig3 | ggfig4)
print(ggfig)
```

```{r dangle length distribution, fig.cap = "dangle length distribution", echo = FALSE, message = FALSE, warning = FALSE}
indel_tsv |>
  pivot_longer(cols = c(uplen, randlen, downlen), names_to = "unmapped", values_to = "unmapped_length") |>
  ggplot(aes(unmapped_length, weight = count, fill = unmapped)) +
  geom_bar() +
  scale_fill_discrete(limits = c("uplen", "randlen", "downlen"), labels = c("updangle length", "random insertion length", "downdangle length")) +
  scale_y_continuous(expand = c(0,0)) ->
  ggfig
print(ggfig)
```

```{r score distribution, fig.cap = "score distribution", echo = FALSE, message = FALSE, warning = FALSE}
indel_tsv |>
  ggplot(aes(score, weight = count)) +
  stat_bin(aes(y = after_stat(density))) +
  geom_density(aes(y = after_stat(density)), color = "red") +
  scale_x_continuous(expand = c(0,0)) +
  scale_y_continuous(name = "density", expand = c(0,0)) ->
  ggfig
print(ggfig)
```

```{r seqLogo, fig.cap = "seqLogo", fig.asp = 0.5, echo = FALSE, message = FALSE, warning = FALSE}
basefreq <- py$basefreq
rownames(basefreq) <- c("A", "C", "G", "T", "D")
ggplot() +
  geom_logo(basefreq, method = "bits", namespace = "ACGTD") +
  scale_x_continuous(breaks = NULL) +
  theme_logo() ->
  ggfig1

ggplot() +
  geom_logo(basefreq, method = "prob", namespace = "ACGTD") +
  scale_x_continuous(breaks = NULL) +
  theme_logo() ->
  ggfig2

ggfig <- ggfig1 / ggfig2
print(ggfig)
```

```{r polygon insertion, fig.cap = "polygon insertion", fig.asp = 0.5, echo = FALSE, message = FALSE, warning = FALSE}
regs <- gregexpr("-+", reflines)
inscounts <- rep(counts, time = unlist(lapply(regs, function(reg){if (reg[1] == -1) return(0); return(length(reg))})))
insposes <- lapply(regs, function(reg) {if (reg[1] == -1) return(NULL); clen <- cumsum(attributes(reg)$match.length); reg - 1 - c(0, clen[seq_len(length(clen)-1)])}) |> unlist()
inslens <- lapply(regs, function(reg) {if (reg[1] == -1) return(NULL); attributes(reg)$match.length}) |> unlist()

tibble(inscount = inscounts, inspos = insposes, inslen = inslens) |>
  group_by(inspos, inslen) |>
  summarise(count = sum(inscount)) |>
  ungroup() ->
  inssumm

polyxs <- rep(inssumm$inspos, each = 4)
polyxs[seq(3, length(polyxs), 4)] <- polyxs[seq(3, length(polyxs), 4)] + inssumm$inslen
polyys <- rep(0, time = length(polyxs))
polyys[2:3 + rep(seq(0, length(polyys) - 1, 4), each = 2)] <- rep(inssumm$count, each = 2)
tibble(polyx = polyxs, polyy = polyys) |>
  ggplot(aes(polyx, polyy)) +
  geom_polygon(color = "black", fill = NA, size = 0.1) +
  scale_x_continuous(name = "pos", expand = c(0,0)) +
  scale_y_continuous(name = "count", expand = c(0,0)) ->
  ggfig
print(ggfig)
```

```{r arc deletion, fig.cap = "arc deletion", fig.asp = 0.5, echo = FALSE, message = FALSE, warning = FALSE}
lposes <- reflines |> strsplit("") |> lapply(function(reg){c(0, cumsum(reg != "-"))})
regs <- gregexpr("-+", querylines)
delcounts <- rep(counts, time = unlist(lapply(regs, function(reg){if (reg[1] == -1) return(0); return(length(reg))})))
delstarts <- sapply(seq_len(length(regs)), function(i){if (regs[[i]][1] == -1) return(NULL); return(lposes[[i]][regs[[i]]])}) |> unlist()
delends <- sapply(seq_len(length(regs)), function(i){if (regs[[i]][1] == -1) return(NULL); return(lposes[[i]][regs[[i]] + attributes(regs[[i]])$match.length])}) |> unlist()

tibble(delcount = delcounts, delstart = delstarts, delend = delends) |>
  group_by(delstart, delend) |>
  summarise(count = sum(delcount)) |>
  ungroup() |>
  ggplot(aes(x0 = (delstart + delend) / 2, y0 = 0, r = (delend - delstart) / 2, start = - pi / 2, end = pi / 2)) +
  geom_arc(aes(size = count), alpha = 0.1) +
  scale_size_continuous(range = c(0.1, 2)) +
  scale_x_continuous(name = "pos", expand = c(0,0)) +
  scale_y_continuous(name = NULL, expand = c(0,0)) ->
  ggfig
print(ggfig)
```
